hypersync-client/src/lib.rs

```diff
@@ -692,6 +694,212 @@ impl Client {
+                if let Some(bearer_token) = &client.bearer_token {
+                    req = req.bearer_auth(bearer_token);
+                }
+                req = req
+                    .header(header::ACCEPT, "text/event-stream")
+                    .timeout(MAX_CONNECTION_AGE);
+
+                match req.send().await {
+                    Ok(res) => {
+                        let status = res.status();
+                        if !status.is_success() {
+                            log::warn!("HTTP error: status code {}", status);
+                            if tx
+                                .send(HeightStreamEvent::Reconnecting {
+                                    delay: reconnect_delay,
+                                })
+                                .await
+                                .is_err()
+                            {
+                                return;
+                            }
+                            tokio::time::sleep(reconnect_delay).await;
+                            reconnect_delay =
+                                std::cmp::min(reconnect_delay * 2, MAX_RECONNECT_DELAY);
+                            continue;
+                        }
+
+                        log::info!("Connected to height SSE stream");
+                        if tx.send(HeightStreamEvent::Connected).await.is_err() {
+                            return;
+                        }
+                        reconnect_delay = INITIAL_RECONNECT_DELAY;
+
+                        let mut byte_stream = res.bytes_stream();
+                        let mut buf = String::new();
+                        let mut connection_active = true;
+
+                        while connection_active {
+                            let next_chunk =
+                                tokio::time::timeout(READ_TIMEOUT, byte_stream.next()).await;
+
+                            match next_chunk {
+                                Err(_) => {
+                                    log::warn!(
+                                        "No data received for {:?}, connection appears dead",
+                                        READ_TIMEOUT
+                                    );
+                                    connection_active = false;
+                                }
+                                Ok(Some(Ok(bytes))) => {
+                                    log::trace!("Received {} bytes from SSE stream", bytes.len());
+
+                                    let chunk_str = String::from_utf8_lossy(&bytes);
+                                    buf.push_str(&chunk_str);
+
+                                    // Process complete SSE messages (delimited by \n\n)
+                                    while let Some(idx) = buf.find("\n\n") {
+                                        let event_block = buf[..idx].to_string();
+                                        buf.drain(..idx + 2);
```

Comment:
I'm a little concerned about how low level and unsafe this could be without breaking it up as a standalone function with unit tests.

I don't know all the lower level details about how chunks of these byte streams could be formed. So I'm hesitant to be confident in this.

Did you consider using a lib with tests like https://github.com/jpopesculian/eventsource-stream

Reply:
@JasoonS perfect. reqwest-eventsource uses eventsource-stream internally. It's just an additional abstraction 👍🏼  perfectly happy with that for now.